src/joystick/hidapi/SDL_hidapi_gip.c

```diff
@@ -0,0 +1,2196 @@
+MAKE_GUID(GUID_Wheel, 0x646979cf, 0x6b71, 0x4e96, 0x8d, 0xf9, 0x59, 0xe3, 0x98, 0xd7, 0x42, 0x0c);
+
+/*
+ * The following GUIDs are observed, but the exact meanings aren't known, so
+ * for now we document them but don't use them anywhere.
+ *
+ * MAKE_GUID(GUID_GamepadEmu, 0xe2e5f1bc, 0xa6e6, 0x41a2, 0x8f, 0x43, 0x33, 0xcf, 0xa2, 0x51, 0x09, 0x81);
+ * MAKE_GUID(GUID_IAudioOnly, 0x92844cd1, 0xf7c8, 0x49ef, 0x97, 0x77, 0x46, 0x7d, 0xa7, 0x08, 0xad, 0x10);
+ * MAKE_GUID(GUID_IControllerProfileModeState, 0xf758dc66, 0x022c, 0x48b8, 0xa4, 0xf6, 0x45, 0x7b, 0xa8, 0x0e, 0x2a, 0x5b);
+ * MAKE_GUID(GUID_ICustomAudio, 0x63fd9cc9, 0x94ee, 0x4b5d, 0x9c, 0x4d, 0x8b, 0x86, 0x4c, 0x14, 0x9c, 0xac);
+ * MAKE_GUID(GUID_IExtendedDeviceFlags, 0x34ad9b1e, 0x36ad, 0x4fb5, 0x8a, 0xc7, 0x17, 0x23, 0x4c, 0x9f, 0x54, 0x6f);
+ * MAKE_GUID(GUID_IHeadset, 0xbc25d1a3, 0xc24e, 0x4992, 0x9d, 0xda, 0xef, 0x4f, 0x12, 0x3e, 0xf5, 0xdc);
+ * MAKE_GUID(GUID_IProgrammableGamepad, 0x31c1034d, 0xb5b7, 0x4551, 0x98, 0x13, 0x87, 0x69, 0xd4, 0xa0, 0xe4, 0xf9);
+ * MAKE_GUID(GUID_IVirtualDevice, 0xdfd26825, 0x110a, 0x4e94, 0xb9, 0x37, 0xb2, 0x7c, 0xe4, 0x7b, 0x25, 0x40);
+ * MAKE_GUID(GUID_OnlineDevAuth, 0x632b1fd1, 0xa3e9, 0x44f9, 0x84, 0x20, 0x5c, 0xe3, 0x44, 0xa0, 0x64, 0x04);
+ */
+
+static const int GIP_DataClassMtu[8] = { 64, 64, 64, 2048, 0, 0, 0, 0 };
+
+typedef struct GIP_Quirks
+{
+    Uint16 vendor_id;
+    Uint16 product_id;
+    Uint32 added_features;
+    Uint32 filtered_features;
+    Uint32 quirks;
+    Uint32 extra_in_system[8];
+    Uint32 extra_out_system[8];
+    GIP_DeviceType device_type;
+} GIP_Quirks;
+
+static const GIP_Quirks quirks[] = {
+    { USB_VENDOR_MICROSOFT, USB_PRODUCT_XBOX_ONE_ELITE_SERIES_1,
+      .added_features = GIP_FEATURE_ELITE_BUTTONS,
+      .filtered_features = GIP_FEATURE_CONSOLE_FUNCTION_MAP },
+
+    { USB_VENDOR_MICROSOFT, USB_PRODUCT_XBOX_ONE_ELITE_SERIES_2,
+      .added_features = GIP_FEATURE_ELITE_BUTTONS | GIP_FEATURE_DYNAMIC_LATENCY_INPUT | GIP_FEATURE_CONSOLE_FUNCTION_MAP },
+
+    { USB_VENDOR_MICROSOFT, USB_PRODUCT_XBOX_SERIES_X,
+      .added_features = GIP_FEATURE_DYNAMIC_LATENCY_INPUT },
+
+    { USB_VENDOR_PDP, USB_PRODUCT_PDP_ROCK_CANDY,
+      .quirks = GIP_QUIRK_NO_HELLO },
+
+    { USB_VENDOR_POWERA, USB_PRODUCT_BDA_XB1_FIGHTPAD,
+      .filtered_features = GIP_FEATURE_MOTOR_CONTROL },
+
+    /*
+     * The controller can attempt to resend the metadata too quickly, but has
+     * bugs handling reliable message handling if things get out of sync.
+     * However, since it just lets us bypass the metadata exchange, let's just
+     * do that instead of having an unreliable init
+     */
+    { USB_VENDOR_POWERA, USB_PRODUCT_BDA_XB1_CLASSIC,
+      .quirks = GIP_QUIRK_BROKEN_METADATA },
+
+    { USB_VENDOR_RAZER, USB_PRODUCT_RAZER_ATROX,
+      .filtered_features = GIP_FEATURE_MOTOR_CONTROL,
+      .device_type = GIP_TYPE_ARCADE_STICK },
+
+    {0},
+};
+
+typedef struct GIP_Header
```

Comment:
Do we need a `#pragma push(pack, 1)`for these?

Reply:
No, the header is decoded, not copied out of the packet. This is why I used bitshifts instead of endianness functions: the data was likely unaligned, and reading unaligned memory is, uh, proscribed on non-x86 architectures. Perhaps GIP_AudioFormat and GUID might, but those structs are designed to be packed properly by default. I'm not sure if that's part of the C spec, however; it might be convention.